Add `JAXEpochIterator` and speed up `get_numpy_iterator`

james77777778 opened the pull request:
There are two major improvements:
1. Utilize shared memory to convert tensor to numpy (`TFDatasetAdapter` and `TorchDataLoaderAdapter`)
2. Introduce `JAXEpochIterator` with prefetching to achieve speed-up
3. Add prefetching to `TorchEpochIterator` to achieve speed-up

The result:
JAX:
|Methods|Training speed||
|-|-|-|
|Current implementation|147ms/step||
|`np.asarray`|137ms/step|-6.8%|
|`np.asarray` + prefetching|128ms/step|-12.3%|

Torch:
|Methods|Training speed||
|-|-|-|
|Current implementation|223ms/step||
|`np.asarray`|214ms/step|-4.0%|
|`np.asarray` + prefetching|203ms/step|-9.0%|

The standalone script (tested with RTX4070):

```python
import numpy as np

# isort: off
import torch  # noqa:F401 dummy import to prevent segmentation fault
import tensorflow as tf

# isort: on


import keras
from keras import backend
from keras import layers
from keras import losses
from keras import optimizers


class TestModel(keras.Model):
    def __init__(self, num_layers=4, **kwargs):
        super().__init__(**kwargs)
        self.num_layers = num_layers
        self.convs = []
        for _ in range(num_layers):
            self.convs.append(
                layers.Conv2D(32, 3, padding="same", activation="relu")
            )
        self.pool = layers.GlobalAveragePooling2D()
        self.fatten = layers.Flatten()
        self.dense = layers.Dense(1)

    def call(self, x):
        for i in range(self.num_layers):
            x = self.convs[i](x)
        x = self.pool(x)
        x = self.fatten(x)
        x = self.dense(x)
        return x


x = np.random.random((1024, 416, 416, 3)).astype("float32")
y = np.random.random((1024, 1)).astype("float32")


def preprocess(x, y):
    x = tf.image.random_brightness(x, 0.2)
    x = tf.image.random_contrast(x, 0.2, 0.5)
    x = tf.image.random_saturation(x, 0.2, 0.5)
    x = tf.image.rgb_to_hsv(x)
    x = tf.image.hsv_to_rgb(x)
    return x, y


train_dataset = tf.data.Dataset.from_tensor_slices((x, y))
train_dataset = train_dataset.batch(32)
train_dataset = train_dataset.map(
    preprocess, num_parallel_calls=tf.data.AUTOTUNE
)
train_dataset = train_dataset.prefetch(tf.data.AUTOTUNE)

model = TestModel(num_layers=4)
model.compile(
    optimizer=optimizers.SGD(learning_rate=0.001),
    loss=losses.MeanSquaredError(),
    jit_compile=False if backend.backend() == "torch" else True,
)
model.fit(train_dataset, epochs=3)

```

Ref:
- https://github.com/google/flax/blob/85dfad242e56098849dbf05e7e4657b3a40820f9/examples/imagenet/train.py#L193
- https://flax.readthedocs.io/en/latest/api_reference/flax.jax_utils.html#flax.jax_utils.prefetch_to_device

fchollet commented on keras/backend/jax/trainer.py:
```diff
@@ -901,3 +888,49 @@ def _purge_model_variables(
+            for data in itertools.islice(numpy_iterator, n):
+                queue.append(_distribute_data(data))
+
+        enqueue(n=2)  # TODO: should we make `n` configurable?
```
A good strategy may be to make this similar to "AUTO" in tf.data, where we gradually increase the value until we hit diminishing returns (or performance reduction) and then revert back to the optimal value.

james77777778 replied:
This sounds like a more user-friendly strategy.

I have some questions about it:
1. Is it acceptable to introduce some level of overhead/complexity to fine-tune `n`?
2. `n` is only configurable at each epoch, and the statistics of one-epoch time might vary with different metrics and callbacks, etc. Is it fine to use one-epoch time to optimize `n`?
3. There is a drawback to setting a large `n`: it will consume a lot of memory on the target device.

References:
- ResNet50 example from Flax uses `2`: https://github.com/google/flax/blob/85dfad242e56098849dbf05e7e4657b3a40820f9/examples/imagenet/train.py#L214
- `torchtnt.utils.data.CudaDataPrefetcher` defaults to `2`: https://pytorch.org/tnt/stable/utils/generated/torchtnt.utils.data.CudaDataPrefetcher.html

fchollet reviewed (commented):
This is great! Thank you for proposing this. Maybe we could something similar for torch as well?

james77777778 commented:
> This is great! Thank you for proposing this. Maybe we could something similar for torch as well?

Sure. I have added a similar prefetching method in `TorchEpochIterator` and replaced `EpochIterator` with it.
You can check the performance gain:
https://github.com/keras-team/keras/pull/18991#issue-2055704197

Note that: the above prefetching only takes effect when it is not a `TorchDataLoaderAdapter`, as Dataloader should have its own prefetching inside.

fchollet reviewed (approved):
LGTM, thank you. Regarding configurability of the number of prefetched samples, we can leave it hardcoded as 2 for now. We could expose it as a `compile()` argument in the future. Implementing autotuning outside tf.data might be overkill.
